internal/engine/planner/struct_matcher_test.go

```diff
@@ -0,0 +1,100 @@
+// Copyright 2021-2025 Zenauth Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+package planner
+
+import (
+	"regexp"
+	"testing"
+	"time"
+
+	"github.com/google/cel-go/parser"
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/require"
+
+	"github.com/cerbos/cerbos/internal/conditions"
+)
+
+func TestStructMatcher(t *testing.T) {
+	tests := []struct {
+		expr string
+		want string
+	}{
+		{
+			expr: "{\"a\": 3}[R.attr.Id] == 4",
+			want: `R.attr.Id == "a" && 4 == 3`,
+		},
+		{
+			expr: "4 == {\"a\": 3}[R.attr.Id]", // can't swap args
+		},
+		{
+			expr: "P.attr.anyMap[R.attr.Id] == R.attr.value",
+		},
+		{
+			expr: `{"a1": {"role": "OWNER"}}[R.id].role == "OWNER"`,
+			want: `R.id == "a1" && "OWNER" == {"role": "OWNER"}.role`,
```

Comment:
If we can determine that `R.id` must be `a1` because there's only one key in the map, can't we apply the same logic to the inner map and realise that it's effectively comparing `"OWNER" == "OWNER"` and eliminate that altogether?

Reply:
This code recognises specific patterns in partially evaluated expressions and transforms them. If transformation happens, the result is going through the partial evaluator again, so I guess this is where this subexpression will be eliminated.